core/src/price_reporter/reporter.rs

```
@@ -70,418 +83,417 @@ pub enum PriceReporterState {
-                } else if price_report == PriceReport::default() {
-                    ExchangeConnectionState::NoDataReported
-                } else {
-                    ExchangeConnectionState::Nominal(price_report)
+                    delay.as_mut().reset(Instant::now() + Duration::from_millis(KEEPALIVE_INTERVAL_MS));
+                },
+
+                // New price streamed from an exchange
+                stream_elem = stream_map.next() => {
+                    if let Some((exchange, res)) = stream_elem {
+                        match res {
+                            Ok(price) => {
+                                let ts = get_current_time();
+                                self.exchange_state
+                                    .new_price(exchange, price, ts);
+
+                                // Stream a price update to the bus
+                                self.config.system_bus.publish(
+                                    price_report_topic_name(&exchange.to_string(), &self.base_token, &self.quote_token),
+                                    SystemBusMessage::PriceReportExchange(PriceReport {
+                                        base_token: self.base_token.clone(),
+                                        quote_token: self.quote_token.clone(),
+                                        exchange: Some(exchange),
+                                        midpoint_price: price,
+                                        local_timestamp: ts,
+                                        reported_timestamp: None
+                                    }),
+                                );
+                            },
+
+                            Err(e) => {
+                                // Restart the connection
+                                log::error!("Error streaming from {exchange}: {e}, restarting connection...");
+                                let new_conn = self.retry_connection(exchange).await.unwrap();
```

I'm not sure if it makes sense to `unwrap` here - the retry can fail, that's what we have `MAX_CONN_RETRIES` for, right? but if it fails (even once), `unwrap`ping here will bring down the connection muxer thread

---

Yeah there's a todo above to handle this case via a `join`, but it's rare enough to not worry about for the moment. Essentially we actually want to propagate these errors